packages/api-client-core/src/operationRunners.ts

```diff
@@ -302,21 +307,69 @@ export const actionRunner: ActionRunner = async <Shape extends RecordShape = any
   }
 };
 
+const processBulkActionResponse = <Shape extends RecordShape = any>(
+  defaultSelection: FieldSelection | null,
+  response: any,
+  records: any,
+  modelSelectionField: string,
+  hasReturnType?: HasReturnType | null
+) => {
+  if (defaultSelection == null) return [];
+  if (!hasReturnType) {
+    return hydrateRecordArray<Shape>(response, records[modelSelectionField]);
+  } else if (typeof hasReturnType == "boolean") {
+    return records.results;
+  } else {
+    return Object.entries(hasReturnType).flatMap(([returnTypeField, innerHasReturnType]) => {
+      const results = records[returnTypeField];
```

Comment:
Similar to my [other comment](https://github.com/gadget-inc/js-clients/pull/536#discussion_r1684984053), `hasReturnType` has this shape, right?

```js
{
  "... on CreateResult": { hasReturnType: false },
  "... on UpdateResult": { hasReturnType: false },
}
```

If so, I wouldn't expect any mapping over `hasReturnType` when it's an object, but rather would expect the code to look a bit more like this:

```js
return records.map((record) => {
  // if typename of record `hasReturnType`, use record as is
  // if typename of record `!hasReturnType`, hydrate
})
```

Basically, what you have on lines 330–339, over the result set in the payload.

Reply:
in the non bulk case it looks like:

``` 
hasReturnType: {
  '... on CreatePostResult': { hasReturnType: true },
  '... on UpdatePostResult': { hasReturnType: false }
};
```

but in this case it looks like:

```
hasReturnType: {
          posts: {
            hasReturnType: {
              '... on Post': { select: true },
              '... on UpsertPostReturnType': { hasReturnType: true }
            }
          }
        };
```

Reply:
Ahhh, I thought it was always the former (wasn't able to keep all of the context in my head 😅). Good to know!